### PR TITLE
Improve pds valid

### PR DESCRIPTION
**Pull request by langbart**
This branch adds the preprocessing step for pds tracks and adds some functions useful to extract some diagnostic properties from pds tracks. Specifically, the function `get_tracks_descriptors` extract some indicators associated with the tracks of each trip aimed to improve the validation step. The preprocessed tracks would consist of a single dataframe populated by diagnostic indicators associated with each trip, then it would be merged with the pds trips table in order to have more information on which perform the validation.

### CODE REVIEW

**Comment by efcaguab** [comment]
Hi @langbart , just wondering how is this going? You reckon it's ready to be merged? Or what else do we need to complete here?

**Comment by langbart** [comment]
Hey @efcaguab I would like to make some more tests on it and work on that next week.

**Comment by langbart** [comment]
### Improve PDS validation

This branch improves the validation of PDS data according to the following steps:

1.  Several diagnostics indicators are extracted from PDS tracks using `get_tracks_descriptors` in `preprocess_pds_tracks`. These diagnostics consist of indicators of the quality associated with each trip' tracks as `outliers_proportion` and `timetrace_dispersion` and other info as the time and distance between consecutive trips.
2. The diagnostics are used in `validate_pds_trips` to improve the detection of potentially unreliable trips and to identify and merge those consecutive trips that are likely to be considered a single trip.

**Comment by langbart** [comment]
### Improve PDS validation

This branch improves the validation of PDS data according to the following steps:

1.  Several diagnostics indicators are extracted from PDS tracks using `get_tracks_descriptors` in `preprocess_pds_tracks`. These diagnostics consist of indicators of the quality associated with each trip' tracks as `outliers_proportion` and `timetrace_dispersion` and other info as the time and distance between consecutive trips.
2. The diagnostics are used in `validate_pds_trips` to improve the detection of potentially unreliable trips and to identify and merge those consecutive trips that are likely to be considered a single trip.

Using these settings, PDS trips decreased from about 77000 to 69000 validated trips.

Known limitations:
The algorithm can identify and merge 2 consecutive trips, I have not yet found a way to include the circumstance (rare) in which there are more than 2 trips representing a potential single trip.

File: R/validate-pds-trips.R [hunk 1]
```diff
@@ -131,3 +205,97 @@ get_preprocessed_trips <- function(pars){
...
+      `Boat Name` = dplyr::first(.data$`Boat Name`),
+      `Boat Gear` = dplyr::first(.data$`Boat Gear`),
+      Community = dplyr::first(.data$Community),
+      `Duration (Seconds)` = sum(.data$`Duration (Seconds)`),
+      `Range (Meters)` = sum(.data$`Range (Meters)`),
+      `Distance (Meters)` = sum(.data$`Distance (Meters)`),
+      IMEI = dplyr::first(.data$IMEI),
+      `Device Id` = dplyr::first(.data$`Device Id`),
+      `Last Seen` = dplyr::last(.data$`Last Seen`),
+      start_end_distance = sum(.data$start_end_distance),
```
Comment (efcaguab) [inline R/validate-pds-trips.R:270]: It probably makes no practical difference, but shouldn't this distance be recalculated for the ones that were merged?

**Comment by efcaguab** [comment]
Looks great @langbart! The only potential problem (I'm not sure it's actually one) is related to the start_end_distance for the merged trips. 

The other thing that I think we should consider is that perhaps we should think about a way to trace down the problematic trips. Not sure how much we need it though but I'm thinking if we need a way keep info about the alerts or wether a trip has been merged? 

Other than that I think it's ready to be merged!

**Review by efcaguab — APPROVED** [review]

**Comment by langbart** [comment]
Hey @efcaguab thanks for the review! You were right, those distances needed to be recalculated.
For what concerns the second point, are you thinking to integrate this pull with Airtable (as we do for trips validation)?

Another thing, I'm thinking to ingest the whole PDS tracks in a single rds file (of around 800 Mb) as it would be useful to generate the tracks' map (or perform FAD-like stuff that Alex was mentioning). Perhaps we could call it from the cloud only when it gets outdated by 5000 trips (just as an example) to synchronize it and update maps and other potential future things, and not for every action. Let me know what do you think about it or have other solutions.

**Comment by efcaguab** [comment]
Hmm I think Airtable integration would be too complicated and probably not needed. Maybe just saving a CSV with those flags or including them as columns in the validated frame would be enough? It's totally OK to not do it for now. Feel free to merge the branch anytime :) 

Re amalgamating all PDS tracks into one file is a great idea. Initially I thought about using BigQuery but your solution is much much simpler and get the job done just as well. Perhaps we can have another github workflow that runs every week or two calling a new function that does that? Note that if we're downloading lots of files from a bucket it might be easier to do it gsutils directly rather than using R. We can then use R to put them all together. Anyways that's another discussion and happy if you can create an issue/PR to do that.

**Pull request merged**